pennylane/templates/state_preparations/mottonen.py

```diff
@@ -134,69 +140,79 @@ def _uniform_rotation_dagger(gate, alpha, control_wires, target_wire):
     ]
 
     for i, control_index in enumerate(control_indices):
-        gate(theta[i], wires=[target_wire])
+        if theta[i] != 0.0:
+            gate(theta[i], wires=[target_wire])
         qml.CNOT(wires=[control_wires[control_index], target_wire])
 
 
 def _get_alpha_z(omega, n, k):
-    r"""Computes the rotation angles for the Z rotations applied to the k'th qubit.
+    r"""Computes the rotation angles required to implement the uniformly-controlled Z rotation
+    applied to the :math:`k`th qubit.
+
+    The :math:`j`th angle is related to the phases omega of the desired amplitudes via:
 
     .. math:: \alpha^{z,k}_j = \sum_{l=1}^{2^{k-1}} \frac{\omega_{(2j-1) 2^{k-1}+l} - \omega_{(2j-2) 2^{k-1}+l}}{2^{k-1}}
 
     Args:
-        omega (float): phases of the input
-        n (int): total number of qubits
+        omega (array): phases of the state to prepare
+        n (int): total number of qubits for the uniformly-controlled rotation
         k (int): index of current qubit
 
     Returns:
         array representing :math:`\alpha^{z,k}`
     """
-    alpha_z_k = np.zeros((2 ** (n - k),), dtype=np.float64)
+    indices1 = [
+        [(2 * j - 1) * 2 ** (k - 1) + l - 1 for l in range(1, 2 ** (k - 1) + 1)]
+        for j in range(1, 2 ** (n - k) + 1)
+    ]
+    indices2 = [
+        [(2 * j - 2) * 2 ** (k - 1) + l - 1 for l in range(1, 2 ** (k - 1) + 1)]
+        for j in range(1, 2 ** (n - k) + 1)
+    ]
 
-    for i in range(len(omega)):
-        j = int(np.ceil((i + 1) * 2 ** (-k)))
-        s_condition = 2 ** (k - 1) * (2 * j - 1)
-        s_i = 1.0 if (i + 1) > s_condition else -1.0
-        alpha_z_k[j - 1] = alpha_z_k[j - 1] + s_i * omega[i] / 2 ** (k - 1)
+    term1 = np.take(omega, indices=indices1)
+    term2 = np.take(omega, indices=indices2)
+    diff = (term1 - term2) / 2 ** (k - 1)
 
-    return alpha_z_k
+    return np.sum(diff, axis=1)
 
 
 def _get_alpha_y(a, n, k):
-    r"""Computes the rotation angles for the Y rotations applied to the k'th qubit.
+    r"""Computes the rotation angles required to implement the uniformly controlled Y rotation
+    applied to the :math:`k`th qubit.
 
-    The angles are related to a as:
+    The :math:`j`th angle is related to the absolute values a of the desired amplitudes via:
```

Comment:
```suggestion
    The :math:`j`th angle is related to the absolute values, a, of the desired amplitudes via:
```
for better readability?